File: admin_app/admin_flutter/lib/models/service_contract.dart
class ServiceContract {
  const ServiceContract({
    required this.id,
    required this.clientId,
    required this.description,
    required this.valuePerPerson,
    required this.defaultPeopleQuantity,
    required this.billingPeriodicity,
    required this.startDate,
    required this.status,
    required this.active,
    required this.rules,
    required this.consumptionItems,
    this.number,
    this.clientName,
    this.notes,
  });

  final int id;
  final String? number;
  final int clientId;
  final String? clientName;
  final String description;
  final double valuePerPerson;
  final double defaultPeopleQuantity;
  final String billingPeriodicity;
  final String startDate;
  final String status;
  final bool active;
  final String? notes;
  final List<ServiceContractRule> rules;
  final List<ServiceContractConsumptionItem> consumptionItems;

  factory ServiceContract.fromJson(Map<String, dynamic> json) {
    return ServiceContract(
      id: json['id'] as int,
      number: json['number'] as String?,
      clientId: json['client_id'] as int,
      clientName: json['client_name'] as String?,
      description: json['description'] as String,
      valuePerPerson: _toDouble(json['value_per_person']),
      defaultPeopleQuantity: _toDouble(json['default_people_quantity']),
      billingPeriodicity: json['billing_periodicity'] as String? ?? 'quinzenal',
      startDate: json['start_date']?.toString() ?? '',
      status: json['status'] as String? ?? 'active',
      active: json['active'] as bool? ?? true,
      notes: json['notes'] as String?,
      rules: (json['rules'] as List<dynamic>? ?? const [])
          .map((item) => ServiceContractRule.fromJson(item as Map<String, dynamic>))
          .toList(),
      consumptionItems:
          (json['consumption_items'] as List<dynamic>? ?? const [])
              .map(
                (item) => ServiceContractConsumptionItem.fromJson(
                  item as Map<String, dynamic>,
                ),
              )
              .toList(),
    );
  }
}

class ServiceContractRule {
  const ServiceContractRule({
    required this.dayType,
    required this.attends,
    required this.charges,
    required this.multiplier,
    this.id,
    this.notes,
  });

  final int? id;
  final String dayType;
  final bool attends;
  final bool charges;
  final double multiplier;
  final String? notes;

  factory ServiceContractRule.fromJson(Map<String, dynamic> json) {
    return ServiceContractRule(
      id: json['id'] as int?,
      dayType: json['day_type'] as String,
      attends: json['attends'] as bool? ?? true,
      charges: json['charges'] as bool? ?? true,
      multiplier: _toDouble(json['multiplier']),
      notes: json['notes'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'day_type': dayType,
      'attends': attends,
      'charges': charges,
      'multiplier': multiplier,
      'notes': _emptyToNull(notes),
    };
  }
}

class ServiceContractConsumptionItem {
  const ServiceContractConsumptionItem({
    required this.productId,
    required this.quantityPerPerson,
    required this.unit,
    required this.wastePercent,
    required this.active,
    this.id,
    this.productName,
    this.notes,
  });

  final int? id;
  final int productId;
  final String? productName;
  final double quantityPerPerson;
  final String unit;
  final double wastePercent;
  final bool active;
  final String? notes;

  factory ServiceContractConsumptionItem.fromJson(Map<String, dynamic> json) {
    return ServiceContractConsumptionItem(
      id: json['id'] as int?,
      productId: json['product_id'] as int,
      productName: json['product_name'] as String?,
      quantityPerPerson: _toDouble(json['quantity_per_person']),
      unit: json['unit'] as String? ?? 'un',
      wastePercent: _toDouble(json['waste_percent']),
      active: json['active'] as bool? ?? true,
      notes: json['notes'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'product_id': productId,
      'quantity_per_person': quantityPerPerson,
      'unit': unit,
      'waste_percent': wastePercent,
      'active': active,
      'notes': _emptyToNull(notes),
    };
  }
}

class ServiceAppointment {
  const ServiceAppointment({
    required this.id,
    required this.contractId,
    required this.appointmentDate,
    required this.dayType,
    required this.peopleQuantity,
    required this.valuePerPerson,
    required this.multiplier,
    required this.totalAmount,
    required this.status,
    required this.stockPosted,
    required this.items,
    this.clientName,
    this.notes,
  });

  final int id;
  final int contractId;
  final String? clientName;
  final String appointmentDate;
  final String dayType;
  final double peopleQuantity;
  final double valuePerPerson;
  final double multiplier;
  final double totalAmount;
  final String status;
  final bool stockPosted;
  final String? notes;
  final List<ServiceAppointmentConsumptionItem> items;

  factory ServiceAppointment.fromJson(Map<String, dynamic> json) {
    return ServiceAppointment(
      id: json['id'] as int,
      contractId: json['contract_id'] as int,
      clientName: json['client_name'] as String?,
      appointmentDate: json['appointment_date']?.toString() ?? '',
      dayType: json['day_type'] as String? ?? 'weekday',
      peopleQuantity: _toDouble(json['people_quantity']),
      valuePerPerson: _toDouble(json['value_per_person']),
      multiplier: _toDouble(json['multiplier']),
      totalAmount: _toDouble(json['total_amount']),
      status: json['status'] as String? ?? 'previsto',
      stockPosted: json['stock_posted'] as bool? ?? false,
      notes: json['notes'] as String?,
      items: (json['items'] as List<dynamic>? ?? const [])
          .map((item) => ServiceAppointmentConsumptionItem.fromJson(item as Map<String, dynamic>))
          .toList(),
    );
  }
}

class ServiceAppointmentConsumptionItem {
  const ServiceAppointmentConsumptionItem({
    required this.productId,
    required this.quantityPlanned,
    required this.quantityConfirmed,
    required this.unit,
    this.id,
    this.productName,
    this.notes,
  });

  final int? id;
  final int productId;
  final String? productName;
  final double quantityPlanned;
  final double quantityConfirmed;
  final String unit;
  final String? notes;

  factory ServiceAppointmentConsumptionItem.fromJson(Map<String, dynamic> json) {
    return ServiceAppointmentConsumptionItem(
      id: json['id'] as int?,
      productId: json['product_id'] as int,
      productName: json['product_name'] as String?,
      quantityPlanned: _toDouble(json['quantity_planned']),
      quantityConfirmed: _toDouble(json['quantity_confirmed']),
      unit: json['unit'] as String? ?? 'un',
      notes: json['notes'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'product_id': productId,
      'quantity_planned': quantityPlanned,
      'quantity_confirmed': quantityConfirmed,
      'unit': unit,
      'notes': _emptyToNull(notes),
    };
  }
}

class ServiceBilling {
  const ServiceBilling({
    required this.id,
    required this.contractId,
    required this.periodStart,
    required this.periodEnd,
    required this.totalAmount,
    required this.status,
    required this.generatedAt,
    required this.items,
    this.number,
    this.receivableId,
  });

  final int id;
  final String? number;
  final int contractId;
  final int? receivableId;
  final String periodStart;
  final String periodEnd;
  final double totalAmount;
  final String status;
  final String generatedAt;
  final List<ServiceBillingItem> items;

  factory ServiceBilling.fromJson(Map<String, dynamic> json) {
    return ServiceBilling(
      id: json['id'] as int,
      number: json['number'] as String?,
      contractId: json['contract_id'] as int,
      receivableId: json['receivable_id'] as int?,
      periodStart: json['period_start']?.toString() ?? '',
      periodEnd: json['period_end']?.toString() ?? '',
      totalAmount: _toDouble(json['total_amount']),
      status: json['status'] as String? ?? 'generated',
      generatedAt: json['generated_at']?.toString() ?? '',
      items: (json['items'] as List<dynamic>? ?? const [])
          .map((item) => ServiceBillingItem.fromJson(item as Map<String, dynamic>))
          .toList(),
    );
  }
}

class ServiceBillingItem {
  const ServiceBillingItem({
    required this.id,
    required this.itemDate,
    required this.description,
    required this.peopleQuantity,
    required this.unitPrice,
    required this.multiplier,
    required this.totalAmount,
    this.appointmentId,
  });

  final int id;
  final int? appointmentId;
  final String itemDate;
  final String description;
  final double peopleQuantity;
  final double unitPrice;
  final double multiplier;
  final double totalAmount;

  factory ServiceBillingItem.fromJson(Map<String, dynamic> json) {
    return ServiceBillingItem(
      id: json['id'] as int,
      appointmentId: json['appointment_id'] as int?,
      itemDate: json['item_date']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      peopleQuantity: _toDouble(json['people_quantity']),
      unitPrice: _toDouble(json['unit_price']),
      multiplier: _toDouble(json['multiplier']),
      totalAmount: _toDouble(json['total_amount']),
    );
  }
}

class ServiceContractPayload {
  ServiceContractPayload({
    required this.clientId,
    required this.description,
    required this.valuePerPerson,
    required this.defaultPeopleQuantity,
    required this.startDate,
    required this.active,
    required this.rules,
    required this.consumptionItems,
    this.status = 'active',
    this.notes,
  });

  final int clientId;
  final String description;
  final double valuePerPerson;
  final double defaultPeopleQuantity;
  final String startDate;
  final String status;
  final bool active;
  final String? notes;
  final List<ServiceContractRule> rules;
  final List<ServiceContractConsumptionItem> consumptionItems;

  Map<String, dynamic> toJson() {
    return {
      'client_id': clientId,
      'description': description.trim(),
      'value_per_person': valuePerPerson,
      'default_people_quantity': defaultPeopleQuantity,
      'billing_periodicity': 'quinzenal',
      'start_date': startDate,
      'status': status,
      'active': active,
      'notes': _emptyToNull(notes),
      'rules': rules.map((rule) => rule.toJson()).toList(),
      'consumption_items': consumptionItems.map((item) => item.toJson()).toList(),
    };
  }
}

double _toDouble(Object? value) {
  if (value == null) return 0;
  if (value is num) return value.toDouble();
  return double.tryParse(value.toString()) ?? 0;
}

String? _emptyToNull(String? value) {
  final trimmed = value?.trim();
  if (trimmed == null || trimmed.isEmpty) return null;
  return trimmed;
}

File: backend/app/models/service_contract.py
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Numeric, String, Text, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class ServiceContract(Base):
    __tablename__ = "service_contracts"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    number: Mapped[Optional[str]] = mapped_column(String(40), unique=True, index=True)
    client_id: Mapped[int] = mapped_column(ForeignKey("clients.id", ondelete="RESTRICT"), index=True)
    description: Mapped[str] = mapped_column(String(180), nullable=False)
    value_per_person: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    default_people_quantity: Mapped[Decimal] = mapped_column(Numeric(12, 3), nullable=False)
    billing_periodicity: Mapped[str] = mapped_column(String(20), nullable=False, default="quinzenal")
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="active")
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    notes: Mapped[Optional[str]] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    client = relationship("Client")
    rules = relationship(
        "ServiceContractAttendanceRule",
        back_populates="contract",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    consumption_items = relationship(
        "ServiceContractConsumptionItem",
        back_populates="contract",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    appointments = relationship("ServiceAppointment", back_populates="contract", lazy="selectin")
    billings = relationship("ServiceBilling", back_populates="contract", lazy="selectin")


class ServiceContractAttendanceRule(Base):
    __tablename__ = "service_contract_attendance_rules"
    __table_args__ = (UniqueConstraint("contract_id", "day_type", name="uq_service_contract_day_type"),)

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    contract_id: Mapped[int] = mapped_column(ForeignKey("service_contracts.id", ondelete="CASCADE"), index=True)
    day_type: Mapped[str] = mapped_column(String(20), nullable=False)
    attends: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    charges: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    multiplier: Mapped[Decimal] = mapped_column(Numeric(8, 4), nullable=False, default=Decimal("1"))
    notes: Mapped[Optional[str]] = mapped_column(Text)

    contract = relationship("ServiceContract", back_populates="rules")


class ServiceContractConsumptionItem(Base):
    __tablename__ = "service_contract_consumption_items"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    contract_id: Mapped[int] = mapped_column(ForeignKey("service_contracts.id", ondelete="CASCADE"), index=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="RESTRICT"), index=True)
    quantity_per_person: Mapped[Decimal] = mapped_column(Numeric(12, 4), nullable=False)
    unit: Mapped[str] = mapped_column(String(20), nullable=False, default="un")
    waste_percent: Mapped[Decimal] = mapped_column(Numeric(7, 2), nullable=False, default=Decimal("0"))
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    notes: Mapped[Optional[str]] = mapped_column(Text)

    contract = relationship("ServiceContract", back_populates="consumption_items")
    product = relationship("Product")


class ServiceAppointment(Base):
    __tablename__ = "service_appointments"
    __table_args__ = (UniqueConstraint("contract_id", "appointment_date", name="uq_service_appointment_date"),)

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    contract_id: Mapped[int] = mapped_column(ForeignKey("service_contracts.id", ondelete="CASCADE"), index=True)
    appointment_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    day_type: Mapped[str] = mapped_column(String(20), nullable=False)
    people_quantity: Mapped[Decimal] = mapped_column(Numeric(12, 3), nullable=False)
    value_per_person: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    multiplier: Mapped[Decimal] = mapped_column(Numeric(8, 4), nullable=False, default=Decimal("1"))
    total_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=Decimal("0"))
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="previsto")
    stock_posted: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    stock_posted_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    confirmed_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    notes: Mapped[Optional[str]] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    contract = relationship("ServiceContract", back_populates="appointments")
    items = relationship(
        "ServiceAppointmentConsumptionItem",
        back_populates="appointment",
        cascade="all, delete-orphan",
        lazy="selectin",
    )


class ServiceAppointmentConsumptionItem(Base):
    __tablename__ = "service_appointment_consumption_items"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    appointment_id: Mapped[int] = mapped_column(ForeignKey("service_appointments.id", ondelete="CASCADE"), index=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="RESTRICT"), index=True)
    quantity_planned: Mapped[Decimal] = mapped_column(Numeric(12, 4), nullable=False, default=Decimal("0"))
    quantity_confirmed: Mapped[Decimal] = mapped_column(Numeric(12, 4), nullable=False, default=Decimal("0"))
    unit: Mapped[str] = mapped_column(String(20), nullable=False, default="un")
    unit_cost: Mapped[Optional[Decimal]] = mapped_column(Numeric(12, 4))
    total_cost: Mapped[Optional[Decimal]] = mapped_column(Numeric(12, 2))
    stock_movement_id: Mapped[Optional[int]] = mapped_column(ForeignKey("stock_movements.id", ondelete="SET NULL"))
    notes: Mapped[Optional[str]] = mapped_column(Text)

    appointment = relationship("ServiceAppointment", back_populates="items")
    product = relationship("Product")
    stock_movement = relationship("StockMovement")


class ServiceBilling(Base):
    __tablename__ = "service_billings"
    __table_args__ = (UniqueConstraint("contract_id", "period_start", "period_end", name="uq_service_billing_period"),)

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    number: Mapped[Optional[str]] = mapped_column(String(40), unique=True, index=True)
    contract_id: Mapped[int] = mapped_column(ForeignKey("service_contracts.id", ondelete="CASCADE"), index=True)
    receivable_id: Mapped[Optional[int]] = mapped_column(ForeignKey("receivables.id", ondelete="SET NULL"), index=True)
    period_start: Mapped[date] = mapped_column(Date, nullable=False)
    period_end: Mapped[date] = mapped_column(Date, nullable=False)
    total_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=Decimal("0"))
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="generated")
    generated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    notes: Mapped[Optional[str]] = mapped_column(Text)

    contract = relationship("ServiceContract", back_populates="billings")
    receivable = relationship("Receivable")
    items = relationship(
        "ServiceBillingItem",
        back_populates="billing",
        cascade="all, delete-orphan",
        lazy="selectin",
    )


class ServiceBillingItem(Base):
    __tablename__ = "service_billing_items"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    billing_id: Mapped[int] = mapped_column(ForeignKey("service_billings.id", ondelete="CASCADE"), index=True)
    appointment_id: Mapped[Optional[int]] = mapped_column(ForeignKey("service_appointments.id", ondelete="SET NULL"))
    item_date: Mapped[date] = mapped_column(Date, nullable=False)
    description: Mapped[str] = mapped_column(String(180), nullable=False)
    people_quantity: Mapped[Decimal] = mapped_column(Numeric(12, 3), nullable=False)
    unit_price: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    multiplier: Mapped[Decimal] = mapped_column(Numeric(8, 4), nullable=False)
    total_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)

    billing = relationship("ServiceBilling", back_populates="items")
    appointment = relationship("ServiceAppointment")
